test/utils/test_utils.py

```diff
@@ -65,117 +65,93 @@ def test_validate_controls():
-
-
-@pytest.mark.parametrize(
-    ["regularization", "expected"],
-    [
-        ("l0", np.array([[0, 3, 5]]).T),
-        ("l1", np.array([[0, 0, 2]]).T),
-        ("l2", np.array([[-2 / 7, 3 / 7, 5 / 7]]).T),
-    ],
-)
-def test_get_prox_1d(regularization, expected):
-    data = np.array([[-2, 3, 5]]).T
-    lam = np.array([[3]])
-
+    with pytest.raises(ValueError):
+        reg(data, lam)
     prox = get_prox(regularization)
-    result = prox(data, lam)
-    assert_array_equal(result, expected)
+    with pytest.raises(ValueError):
+        prox(data, lam)
 
 
 @pytest.mark.parametrize(
-    ["regularization", "expected"],
-    [
-        ("l0", np.array([[0, 3, 5], [-7, 11, 0]]).T),
-        ("l1", np.array([[0, 0, 2], [-4, 8, 0]]).T),
-        ("l2", np.array([[-2 / 7, 3 / 7, 5 / 7], [-7 / 7, 11 / 7, 0 / 7]]).T),
-    ],
+    "regularization", ["weighted_l0", "weighted_l1", "weighted_l2"]
 )
-def test_get_prox_2d(regularization, expected):
-    data = np.array([[-2, 3, 5], [-7, 11, 0]]).T
-    lam = np.array([[3]])
-
-    prox = get_prox(regularization)
-    result = prox(data, lam)
-    assert_array_equal(result, expected)
-
-
 @pytest.mark.parametrize(
-    ["regularization", "expected"],
-    [
-        ("l0", np.array([[0, 3, 5]]).T),
-        ("l1", np.array([[0, 1, 4.5]]).T),
-        ("l2", np.array([[-2 / 7, 3 / 5, 5 / 2]]).T),
-    ],
+    "lam", [np.array([[1, 2]]), np.array([1, 2, 3]), np.array([[1, 2, 3]]).T, 1]
 )
-def test_get_weighted_prox_1d(regularization, expected):
-    data = np.array([[-2, 3, 5]]).T
-    lam = np.array([[3, 2, 0.5]]).T
-
+def test_get_weighted_prox_and_regularization_bad_shape(regularization, lam):
+    data = np.array([[-2, 5]]).T
+    reg = get_regularization(regularization)
+    with pytest.raises(ValueError):
+        reg(data, lam)
     prox = get_prox(regularization)
-    result = prox(data, lam)
-    assert_array_equal(result, expected)
+    with pytest.raises(ValueError):
+        prox(data, lam)
 
 
 @pytest.mark.parametrize(
-    ["regularization", "expected"],
+    ["regularization", "lam", "expected"],
     [
-        ("l0", np.array([[0, 3, 5], [-7, 11, 0]]).T),
-        ("l1", np.array([[0, 1, 4.5], [-6, 0, 0]]).T),
-        ("l2", np.array([[-2 / 7, 3 / 5, 5 / 2], [-7 / 3, 11 / 27, 0 / 35]]).T),
+        ("l0", 3, np.array([[0, 5]]).T),
+        ("l1", 3, np.array([[0, 2]]).T),
+        ("l2", 3, np.array([[-2 / 7, 5 / 7]]).T),
+        ("weighted_l0", np.array([[3, 2]]).T, np.array([[0, 5]]).T),
+        ("weighted_l1", np.array([[3, 2]]).T, np.array([[0, 3]]).T),
+        ("weighted_l2", np.array([[3, 2]]).T, np.array([[-2 / 7, 5 / 5]]).T),
     ],
 )
-def test_get_weighted_prox_2d(regularization, expected):
-    data = np.array([[-2, 3, 5], [-7, 11, 0]]).T
-    lam = np.array([[3, 2, 0.5], [1, 13, 17]]).T
+def test_get_prox(regularization, expected, lam):
+    data = np.array([[-2, 5]]).T
 
     prox = get_prox(regularization)
     result = prox(data, lam)
```

Comment:
This we need to think about a little mathematically, so it's easier to see why `expected` is expected.

Reply:
Consider the prox of different regularizers with only x data = 2

With l2 and weight=1, prox should be 1 because its the arg min of x^2 + (2-x)^2

with l0 and weight=1, prox should be 2, because the only way to reduce the regularizer is to set x=0

With l1 and weight=1, it should be (1.5, 0).  Bit trickier because it's the min of (2-x)^2 + x, which requires some simple algebra.